pkg/kubecli/vm.go

```diff
@@ -126,56 +126,78 @@ func (s *BinaryReadWriter) err(err error) error {
 	return err
 }
 
-type RoundTripCallback func(conn *websocket.Conn, resp *http.Response, err error) error
+type RoundTripCallback func(conn *websocket.Conn, resp *http.Response) error
 
 type WebsocketRoundTripper struct {
 	Dialer *websocket.Dialer
-	Do     RoundTripCallback
+	Conn   *websocket.Conn
+	Resp   *http.Response
 }
 
-func (d *WebsocketRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
-	conn, resp, err := d.Dialer.Dial(r.URL.String(), r.Header)
-	if err == nil {
-		defer conn.Close()
+func (d *WebsocketRoundTripper) checkResponse(resp *http.Response) error {
+
+	if resp == nil {
+		return fmt.Errorf("no response received")
 	}
-	return resp, d.Do(conn, resp, err)
-}
 
-type wsCallbackObj struct {
-	in  io.Reader
-	out io.Writer
+	// look for known issues
+	switch resp.StatusCode {
+	case http.StatusNotFound:
+		return fmt.Errorf("Virtual Machine not found.")
+	case http.StatusInternalServerError:
+		return fmt.Errorf("Websocket failed due to internal server error.")
+	}
+
+	return nil
 }
 
-func (obj *wsCallbackObj) WebsocketCallback(ws *websocket.Conn, resp *http.Response, err error) error {
+func (d *WebsocketRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
 
+	conn, resp, err := d.Dialer.Dial(r.URL.String(), r.Header)
 	if err != nil {
 		if resp != nil && resp.StatusCode != http.StatusOK {
 			buf := new(bytes.Buffer)
 			buf.ReadFrom(resp.Body)
-			return fmt.Errorf("Can't connect to websocket (%d): %s\n", resp.StatusCode, buf.String())
+			return resp, fmt.Errorf("Can't connect to websocket (%d): %s\n", resp.StatusCode, buf.String())
 		}
-		return fmt.Errorf("Can't connect to websocket: %s\n", err.Error())
+		return resp, fmt.Errorf("Can't connect to websocket: %s\n", err.Error())
+	}
+
+	// Are we abusing the RoundTrip() abstraction?
+	// (see https://golang.org/pkg/net/http/#RoundTripper)
+	// We *are* "executing a single HTTP transaction, returning
+	// a Response for the provided Request."
+	// Also, it should be ok to consume the body later, in Stream().
```

Comment:
That would only work for non-interactive commands, where you would get the whole body back. A round-tripper encloses the complete request. You will have to keep the round-tripper open.

Reply:
Got it, fixing.